### ISSUE TITLE
심규호_과제방향 설정

### ISSUE THREAD

**Issue by kyuho90** [Epheus/AI-X#3]
#심규호

# 방향 1.거래 트렌드 분석: 특정 지역에서의 부동산 거래 트렌드(예: 가격, 면적, 건축 연도별 변화 등)를 분석하여 미래의 가격 예측
# 방향 2.이상 거래 탐지: 부동산 거래 데이터에서 이상치(outlier)를 탐지하여 사기성 거래나 비정상적인 패턴을 찾아내는 모델 개발 (옵션)

# 방향 1. 분석 목표/ 방향
# 랜덤 포레스트(Random Forest): 주택 가격 예측에 적합한 회귀 모델 구성 
  PCA(주성분 분석): 고차원 데이터를 저차원으로 축소하여 주요 변수를 추출하고, 데이터의 분산 찾기
  t-SNE: 데이터 시각화를 위해 차원을 축소하여 특정 지역 내의 거래 트렌드를 시각적으로 분석

# 코드 예시 : import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
import matplotlib.pyplot as plt

# 샘플 데이터 로드 (실제 데이터로 대체)
data = pd.read_csv('real_estate_data.csv')

# 데이터 전처리: 결측치 제거 및 특성 선택 (예: price, area, year_built 등)
X = data[['price', 'area', 'year_built']]
y = data['price']

# 학습 데이터와 테스트 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 랜덤 포레스트 모델 학습
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# 예측 및 평가
predictions = model.predict(X_test)
print(f'MAE: {np.mean(np.abs(predictions - y_test))}')

# PCA를 이용한 차원 축소 및 시각화
pca = PCA(n_components=2)
X_pca = pca.fit_transform(X)
plt.scatter(X_pca[:, 0], X_pca[:, 1], c=y, cmap='viridis')
plt.xlabel('Principal Component 1')
plt.ylabel('Principal Component 2')
plt.title('PCA of Real Estate Data')
plt.colorbar()
plt.show()

# t-SNE를 이용한 시각화
tsne = TSNE(n_components=2, perplexity=40)
X_tsne = tsne.fit_transform(X)
plt.scatter(X_tsne[:, 0], X_tsne[:, 1], c=y, cmap='viridis')
plt.xlabel('t-SNE Component 1')
plt.ylabel('t-SNE Component 2')
plt.title('t-SNE of Real Estate Data')
plt.colorbar()
plt.show()

##############방향1.추가 - Neural Network를 사용하여 특정 변수의 중요도(w값)를 찾고, 연도별 데이터를 바탕으로 모델의 적합성을 평가하는 방법을 설명하겠습니다. 이 과정에서는 LSTM을 사용하여 시계열 데이터를 분석하고, SHAP(SHapley Additive exPlanations)를 사용하여 각 변수의 중요도를 평가
pip install numpy pandas tensorflow shap
PYTHON
Copy
import numpy as np
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense
import shap

# 1. 데이터 준비
# 예제 데이터프레임 (실제 데이터로 대체 필요)
data = {
    '연도': [2020, 2021, 2022, 2023, 2024],
    '지역': ['서울', '부산', '대구', '인천'],
    '평균_가격': [500, 300, 250, 400, 550, 350, 280, 420, 600, 380]
}
df = pd.DataFrame(data)

# 연도별 지역별 평균 가격 계산
df_pivot = df.pivot_table(index='연도', columns='지역', values='평균_가격').fillna(0)

# 2. 데이터 정규화
scaler = MinMaxScaler()
scaled_data = scaler.fit_transform(df_pivot)

# 시계열 데이터 생성 (2024년을 기반으로 2025년 예측)
def create_dataset(data, time_step=1):
    X, y = [], []
    for i in range(len(data)-time_step-1):
        a = data[i:(i+time_step), :]
        X.append(a)
        y.append(data[i + time_step, :])
    return np.array(X), np.array(y)

time_step = 1
X, y = create_dataset(scaled_data, time_step)

# 학습 데이터와 테스트 데이터 분리 (2024년 데이터로 학습, 2025년 데이터 예측)
train_size = int(len(X) * 0.8)
X_train, X_test = X[:train_size], X[train_size:]
y_train, y_test = y[:train_size], y[train_size:]

# 3. LSTM 모델 구성
model = Sequential()
model.add(LSTM(50, return_sequences=True, input_shape=(time_step, X_train.shape,[object Object],)))
model.add(LSTM(50))
model.add(Dense(X_train.shape,[object Object],))
model.compile(optimizer='adam', loss='mean_squared_error')

# 모델 학습
model.fit(X_train, y_train, epochs=100, batch_size=64, validation_data=(X_test, y_test))

# 4. 2025년 데이터 예측
last_year_data = scaled_data[-time_step:]
next_year_prediction = []
for _ in range(12):  # 2025년 한 해의 데이터를 예측 (월별 예측 가정)
    predicted_price = model.predict(last_year_data.reshape(1, time_step, -1))
    next_year_prediction.append(predicted_price,[object Object],)
    last_year_data = np.append(last_year_data[1:], predicted_price, axis=0)

# 예측값 역정규화
next_year_prediction = scaler.inverse_transform(np.array(next_year_prediction).reshape(1, -1, X_train.shape,[object Object],))
print("2025년 지역별 평균 가격 예측:")
print(next_year_prediction)

# 5. SHAP를 사용한 변수 중요도 분석
# SHAP Explainer 생성
explainer = shap.KernelExplainer(model.predict, X_train)
shap_values = explainer.shap_values(X_test)

# SHAP 값 시각화
shap.summary_plot(shap_values, X_test, feature_names=df_pivot.columns)

# 6. 모델 적합성 평가
# RMSE 계산
predictions = model.predict(X_test)
predictions = scaler.inverse_transform(predictions)
y_test_inv = scaler.inverse_transform(y_test)
rmse = np.sqrt(np.mean((predictions - y_test_inv) ** 2))
print(f'RMSE: {rmse}')

데이터 준비 및 전처리: 데이터를 연도별로 분리하고, 각 연도별로 지역별 평균 가격을 계산합니다. 데이터를 정규화하여 모델의 학습 성능을 높입니다.

LSTM 모델 구성 및 학습: LSTM을 사용하여 시계열 데이터를 학습합니다.

SHAP를 사용한 변수 중요도 분석: SHAP를 사용하여 각 변수의 중요도를 평가합니다. shap.summary_plot을 통해 변수의 중요도를 시각화합니다.

모델 적합성 평가: RMSE를 계산하여 모델의 예측 성능을 평가합니다.

이 코드를 통해 LSTM을 사용하여 시계열 데이터를 분석하고, SHAP를 사용하여 각 변수의 중요도를 평가할 수 있습니다. 이를 통해 2025년 지역별 부동산 가격을 예측하고, 모델의 적합성을 평가할 수 있습니다.

# 방향 2.이상 거래 탐지: 부동산 거래 데이터에서 이상치(outlier)를 탐지하여 사기성 거래나 비정상적인 패턴을 찾아내는 모델 개발 (옵션)

# 방향 2. 분석 목표/ 방향

랜덤 포레스트: 변수 중요도를 통해 이상 거래 패턴을 탐지
LDA(선형 판별 분석): 고차원 데이터의 패턴을 학습하여 정상 거래와 이상 거래를 구분
PCA와 t-SNE: 데이터를 시각화하여 이상치를 시각적으로 식별

## 코드 예시
from sklearn.ensemble import RandomForestClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
from sklearn.metrics import classification_report

# 이상 거래 탐지 모델 (정상 거래와 이상 거래로 레이블링된 데이터 필요)
X_train, X_test, y_train, y_test = train_test_split(X, y_labels, test_size=0.2, random_state=42)

# 랜덤 포레스트 분류기 학습 (정상 거래: 0, 이상 거래: 1로 레이블링)
clf = RandomForestClassifier(n_estimators=100, random_state=42)
clf.fit(X_train, y_train)

# 예측 및 평가
y_pred = clf.predict(X_test)
print(classification_report(y_test, y_pred))

# LDA를 이용한 차원 축소 및 시각화
lda = LDA()
X_lda = lda.fit_transform(X, y)
plt.scatter(X_lda[:, 0], X_lda[:, 1], c=y, cmap='viridis')
plt.xlabel('LDA Component 1')
plt.ylabel('LDA Component 2')
plt.title('LDA of Real Estate Data')
plt.colorbar()
plt.show()
